ProcessLib/ThermoHydroMechanics/ThermoHydroMechanicsFEM.h

```diff
@@ -0,0 +1,889 @@
+            auto const& N_T = N_p;
+            auto const& dNdx_T = dNdx_p;
+            auto T_int_pt = N_T * T;
+            auto p_int_pt = N_T * p;
+
+            auto const x_coord =
+                interpolateXCoordinate<ShapeFunctionDisplacement,
+                                       ShapeMatricesTypeDisplacement>(_element,
+                                                                      N_u);
+
+            auto const B = LinearBMatrix::computeBMatrix<
+                DisplacementDim, ShapeFunctionDisplacement::NPOINTS,
+                typename BMatricesType::BMatrixType>(dNdx_u, N_u, x_coord,
+                                                     _is_axially_symmetric);
+
+            auto& eps = _ip_data[ip].eps;
+            auto const& sigma_eff = _ip_data[ip].sigma_eff;
+
+            double const S = _process_data.specific_storage(t, x_position)[0];
+            double const K_over_mu =
+                _process_data.intrinsic_permeability(t, x_position)[0] /
+                _process_data.getFluidViscosity(p_int_pt, T_int_pt);
+            double const alpha_s =
+                _process_data.solid_linear_thermal_expansion_coefficient(
+                    t, x_position)[0];
+            double const beta_f =
+                _process_data.fluid_volumetric_thermal_expansion_coefficient(
+                    t, x_position)[0];
+            double const lambda_f =
+                _process_data.getFluidThermalConductivity(p_int_pt, T_int_pt);
+            double const lambda_s =
+                _process_data.solid_thermal_conductivity(t, x_position)[0];
+            double const C_f =
+                _process_data.getFluidHeatCapacity(p_int_pt, T_int_pt);
+            double const C_s =
+                _process_data.solid_specific_heat_capacity(t, x_position)[0];
+            double const T0 =
+                _process_data.reference_temperature(t, x_position)[0];
+            auto const alpha = _process_data.biot_coefficient(t, x_position)[0];
+            auto rho_sr = _process_data.solid_density(t, x_position)[0];
+            auto rho_f = _process_data.getFluidDensity(p_int_pt, T_int_pt);
+            auto const porosity = _process_data.porosity(t, x_position)[0];
+            auto const& b = _process_data.specific_body_force;
+
+            // calculate linear thermal strain
+            // assume isotropic thermal expansion
+
+            double delta_T(T_int_pt - T0);
+            double const thermal_strain = alpha_s * delta_T;
+
+            double rho_s = rho_sr * (1 - 3 * thermal_strain);
+
+            auto velocity = (-K_over_mu * dNdx_p * p).eval();
+            velocity += K_over_mu * rho_f * b;
+
+            auto const& identity2 = MaterialLib::SolidModels::Invariants<
+                KelvinVectorDimensions<DisplacementDim>::value>::identity2;
+
+            //
+            // displacement equation, displacement part (K_uu)
+            //
+            eps.noalias() = B * u;
+
```

Comment:
Remove this blank line since the calculation of eps, which is actuarially the strain increment,   is for updateConstitutiveRelation.

Reply:
fixed